### PR TITLE
fix: Fix block/chunk production prometheus-metrics for kicked out validators

### PR DESCRIPTION
**Pull request by tayfunelmas**
Related [thread](https://near.zulipchat.com/#narrow/stream/407237-core.2Fstateless-validation/topic/Non-validator.20node.20expected.20to.20produce.20chunks/near/456509732)

We update the prometheus metrics for block/chunk production of validators by looping over the shards to record expected/produced_by_shard metric here: https://github.com/near/nearcore/blob/aaff26cfe0ababaa85ad259e701f07c7f797c79c/chain/client/src/info.rs#L506

But if a validator is kicked out in the previous epoch, we need to remove the metrics (to keep only the metrics for the active validators in the current epoch). In this case, the current code records an empty shard id list for the kicked-out validator here: https://github.com/near/nearcore/blob/aaff26cfe0ababaa85ad259e701f07c7f797c79c/chain/client/src/info.rs#L861

This prevents the loop to iterate over the shards and update the metric per shard (while the metrics for aggregated expected/produced are updated to 0).

To fix this, we explicitly create an enum called `ValidatorProductionStatus` which indicates whether the validator is active (`ValidatorProductionStatus::Validator`) or kicked out (`ValidatorProductionStatus::Kickout`) in the current epoch. The old code indicates this with the same struct `ValidatorProductionStats` with 0 values and empty vectors. Then for the kicked out validators, the code loops over the shard ids in the current epoch to remove the metrics for the validator.

### CODE REVIEW

**Comment by tayfunelmas** [comment]
Question to @andrei-near: Should we delete the label values instead of setting them to 0? Which one would make more sense from the Grafana monitoring perspective?

**Comment by andrei-near** [comment]
> Question to @andrei-near: Should we delete the label values instead of setting them to 0? Which one would make more sense from the Grafana monitoring perspective?
> 
> UPDATE: Now updated the PR to clear the metrics instead of setting them to 0.

Yes, removing them would be the right way. Otherwise we can end up with a list of historical accounts

**Comment by tayfunelmas** [comment]
> Yes, removing them would be the right way. Otherwise we can end up with a list of historical accounts

Ok, this PR reflects this decision now.

File: chain/client/src/info.rs [hunk 1]
```diff
@@ -142,10 +142,14 @@ impl InfoHelper {
...
 
     /// Count which shards are tracked by the node in the epoch indicated by head parameter.
-    fn record_tracked_shards(head: &Tip, client: &crate::client::Client) {
-        let validator_signer = client.validator_signer.get();
-        let me = validator_signer.as_ref().map(|x| x.validator_id());
-        if let Ok(shard_ids) = client.epoch_manager.shard_ids(&head.epoch_id) {
+    fn record_tracked_shards(
+        head: &Tip,
+        client: &crate::client::Client,
+        shard_ids: Option<Vec<ShardId>>,
```
Comment (pugachAG) [inline chain/client/src/info.rs:148]: nit: to me it looks a bit weird to see a function that accepts an optional arg and does nothing if it is `None`, I think it would be cleaner to make it `shard_ids: &[ShardId]` and check for `None` on the caller side

File: chain/client/src/info.rs [hunk 2]
```diff
@@ -142,10 +142,14 @@ impl InfoHelper {
     }
 
     /// Count which shards are tracked by the node in the epoch indicated by head parameter.
```
Comment (wacban) [inline chain/client/src/info.rs:144]: The comment seems totally unrelated to what this method does. Can you fix it?

File: chain/client/src/info.rs (hunk 1 above)
Comment (wacban) [reply chain/client/src/info.rs:148]: +1 And it also feels like this should always be the list of all shard ids at the given epoch. Assuming this is the case it would be better to pass the shard layout.

File: chain/client/src/info.rs [hunk 3]
```diff
@@ -877,23 +917,23 @@ impl ValidatorProductionStats {
...
-fn get_validator_epoch_stats(
+/// Converts EpochValidatorInfo into a vector of ValidatorProductionStatus.
+fn get_validator_production_status(
     current_validator_epoch_info: EpochValidatorInfo,
-) -> Vec<ValidatorProductionStats> {
-    let mut stats = vec![];
+) -> Vec<ValidatorProductionStatus> {
+    let mut status = vec![];
     // Record kickouts to replace latest stats of kicked out validators with zeros.
     for kickout in current_validator_epoch_info.prev_epoch_kickout {
```
Comment (wacban) [inline chain/client/src/info.rs:930]: sanity check - Should that be prev_epoch or prev_prev_epoch kickouts? Does this work as following?
* Validator is bad in epoch T
* It is removed from the validator list of epoch T + 2 
* Is its status kickout in epoch T + 1 ? It's not on chain right? Is this info derived from the validator sets?

File: chain/client/src/info.rs (hunk 1 above)
Comment (tayfunelmas) [reply chain/client/src/info.rs:148]: yeah, changing it to ShardLayout, since there is also a problem with the chunk validator stats using a subset of the shard ids, so using ShardLayout should be cleaner than a vector of shard ids.

File: chain/client/src/info.rs [hunk 4]
```diff
@@ -485,38 +497,67 @@ impl InfoHelper {
     }
 
     /// Updates the prometheus metrics to track the block and chunk production and endorsement by validators.
-    fn update_validator_metrics(validator_epoch_stats: Vec<ValidatorProductionStats>) {
+    fn update_validator_metrics(
+        validator_status: Vec<ValidatorProductionStatus>,
+        shard_ids: Option<Vec<ShardId>>,
+    ) {
+        let shards = shard_ids.unwrap_or(vec![]);
```
[the remark this answers is not in the document]
Comment (tayfunelmas) [reply chain/client/src/info.rs:504]: done.

File: chain/client/src/info.rs (hunk 2 above)
Comment (tayfunelmas) [reply chain/client/src/info.rs:144]: done

File: chain/client/src/info.rs (hunk 3 above)
Comment (tayfunelmas) [reply chain/client/src/info.rs:930]: Good point. See this code: https://github.com/near/nearcore/blob/3bdc50d2a1c537bd300ce0c0df2813cd574c3ee9/chain/epoch-manager/src/lib.rs#L1563

So the prev_epoch_kickout (of T+1) is computed from the next epoch info (T+2), and the kickouts are referring to the validator stats from T.

Comment (wacban) [reply chain/client/src/info.rs:930]: An alternative approach would be to forget about kickouts and just sweep all the currently set metrics and unset them and then just set the metrics for current validators. Not sure about perf so maybe do it only at the first block of epoch.

**Comment by tayfunelmas** [comment]
I will go ahead and merge this PR. I will follow-up with other fixes (per-shard endorsement metrics and prev-prev kickout discussion) in upcoming PRs.
